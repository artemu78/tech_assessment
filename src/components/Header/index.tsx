import { useState } from "react";
import IconButton from "@mui/material/IconButton";
import Dialog from "@mui/material/Dialog";
import DialogActions from "@mui/material/DialogActions";
import DialogContent from "@mui/material/DialogContent";
import MenuIcon from "@mui/icons-material/Menu";
import Menu from "@mui/material/Menu";
import MenuItem from "@mui/material/MenuItem";
import Button from "@mui/material/Button";
import styles from "./styles.module.css";

const Header = (): JSX.Element => {
  const [isModalOpen, setModalOpen] = useState(false);
  const [anchorEl, setMenuAnchorEl] = useState<null | HTMLElement>(null);

  const handleMenuClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    console.log("burger menu click");
    setMenuAnchorEl(event.currentTarget);
  };

  const open = Boolean(anchorEl);

  const handleDialogClose = () => {
    console.log("close dialog");
    setModalOpen(false);
  };
  const handleMenuClose = () => {
    console.log("close burger menu");
    setMenuAnchorEl(null);
  };

  const aboutItemClick = () => {
    console.log("open dialog");
    setModalOpen(true);
    setMenuAnchorEl(null);
  };

  console.log({ open, isModalOpen });
  return (
    <header>
      <a href="https://war.ukraine.ua/support-ukraine/" target="_blank" rel="noreferrer">
        <div className={styles.praporBlue} />
        <div className={styles.praporYellow} />
      </a>

      <div className={styles.titleLine}>
        <IconButton
          size="large"
          edge="start"
          color="inherit"
          aria-label="menu"
          sx={{ mr: 2, p: 2 }}
          aria-controls={open ? "basic-menu" : undefined}
          aria-haspopup="true"
          aria-expanded={open ? "true" : undefined}
          onClick={handleMenuClick}
          data-testid="header_menuButton"
        >
          <MenuIcon />
        </IconButton>
        <h1>Technical Assessment</h1>
      </div>
      <Dialog
        onClose={() => {
          console.log("try to close dialog");
          handleDialogClose();
        }}
        open={isModalOpen}
        data-testid="header_aboutDialog"
      >
        <DialogContent>
          This is open source technical questions assessment.
          <br />
          <br />
          Choose theme. Answer questions one by one or choose one you feel comfortable with.
          <br />
          You can find your right and wrong answers counted.
          <br />
          <br />
          Questions and answers - are open source community baked{" "}
          <a
            href="https://github.com/ebazhanov/linkedin-skill-assessments-quizzes/"
            target="_blank"
            rel="noreferrer"
          >
            repository
          </a>
          . You see a fresh version every time you load page.
          <br />
          Feel free to{" "}
          <a
            href="https://docs.github.com/en/get-started/exploring-projects-on-github/finding-ways-to-contribute-to-open-source-on-github"
            target="_blank"
            rel="noreferrer"
          >
            contribute
          </a>
        </DialogContent>
        <DialogActions>
          <Button onClick={handleDialogClose} data-testid="header-closeButton">
            Close
          </Button>
        </DialogActions>
      </Dialog>
      <Menu
        id="basic-menu"
        anchorEl={anchorEl}
        open={open}
        onClose={handleMenuClose}
        MenuListProps={{
          "aria-labelledby": "basic-button",
        }}
      >
        <MenuItem onClick={aboutItemClick} data-testid="header_aboutItem">
          About
        </MenuItem>
      </Menu>
    </header>
  );
};

export default Header;
