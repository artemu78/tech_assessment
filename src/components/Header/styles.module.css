.praporBlue {
  background-color: #0066cc;
  height: 10px;
}

.praporYellow {
  background-color: #ffcc00;
  height: 10px;
}

.titleLine {
  display: flex;
  flex-direction: row;
  align-items: center;
  padding: 0 10px;
}

.titleLine > h1 {
  flex-grow: 1;
}
