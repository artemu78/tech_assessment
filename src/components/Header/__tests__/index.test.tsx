import React from "react";
import {
  render,
  screen,
  fireEvent,
  waitFor,
  waitForElementToBeRemoved,
} from "@testing-library/react";
const { default: userEvent } = require("@testing-library/user-event");
import Dialog from "@mui/material/Dialog";
import Header from "..";

// jest.mock("../foo-bar-baz", () => {
//   const originalModule = jest.requireActual("../foo-bar-baz");
//   return {
//     __esModule: true,
//     ...originalModule,
//     onClose: jest.fn(() => console.log()),
//   };
// });

describe("Header", () => {
  test("render snapshot", () => {
    const { container } = render(<Header />);
    expect(container).toMatchSnapshot();
  });
  test("hide and show dialog", async () => {
    // const customSetState = (initial) => [initial, (val) => {}];

    // const setModalOpen = jest.fn();
    // const setMenuAnchorEl = jest.fn();

    // const setModalOpen = jest.fn().mockReturnValueOnce(false);

    // const setstate = <T,>(init: T) => {
    //   const hook = { val: init };
    //   return [hook.val, (newval: any) => (hook.val = newval)];
    // };

    // React.useState = jest
    //   .fn()
    //   .mockReturnValueOnce(setstate(false))
    //   .mockReturnValueOnce(setstate(null));

    const { container, getByTestId } = render(<Header />);
    expect(screen.getByRole("link")).toHaveAttribute(
      "href",
      "https://war.ukraine.ua/support-ukraine/"
    );
    expect(screen.queryByTestId("header_menuButton")).toBeInTheDocument();
    expect(screen.queryByTestId("header_aboutItem")).not.toBeInTheDocument();

    fireEvent.click(getByTestId("header_menuButton")); // 1st click on burger button on the left
    expect(screen.queryByTestId("header_aboutItem")).toBeInTheDocument();
    await waitFor(() => screen.getByTestId("header_menuButton"));
    await waitFor(() => screen.getByTestId("header_aboutItem"));
    expect(screen.queryByTestId("header_aboutDialog")).not.toBeInTheDocument();
    expect(screen.queryByTestId("header_aboutItem")).toBeVisible();

    console.log("before 2nd click", new Date().toISOString());
    await userEvent.click(getByTestId("header_menuButton"), { delay: 1000 }); // 2nd click on burger button on the left
    console.log("after 2nd click", new Date().toISOString());

    expect(screen.queryByTestId("header_aboutItem")).toBeInTheDocument();
    expect(screen.queryByTestId("header_aboutItem")).toBeVisible();

    await waitFor(() => screen.getByTestId("header_menuButton"));
    await waitFor(() => screen.getByTestId("header_aboutItem"));
    expect(screen.queryByTestId("header_aboutDialog")).not.toBeInTheDocument();

    fireEvent.click(getByTestId("header_aboutItem")); // click on "About" menu item
    await waitFor(() => screen.getByTestId("header_aboutDialog")); // wait until dialog appear
    expect(screen.queryByTestId("header_aboutDialog")).toBeInTheDocument();

    expect(
      screen
        .getAllByRole("link")
        .find(
          (el) =>
            (el as HTMLLinkElement).href ===
            "https://docs.github.com/en/get-started/exploring-projects-on-github/finding-ways-to-contribute-to-open-source-on-github"
        )
    ).toBeInTheDocument();
    expect(
      screen
        .getAllByRole("link")
        .find(
          (el) =>
            (el as HTMLLinkElement).href ===
            "https://github.com/ebazhanov/linkedin-skill-assessments-quizzes/"
        )
    ).toBeInTheDocument();

    // fireEvent.click(getByTestId("header_menuButton")); // click on burger button on the left
    // // fireEvent.click(document.body);

    // await fireEvent.keyDown(document.body, {
    //   key: "Escape",
    //   keyCode: 27,
    //   which: 27,
    //   delay: 1000,
    // });

    // await userEvent.click(document.body, { delay: 1000 });
    // await userEvent.click(getByTestId("header-closeButton"), { delay: 1000 });

    // // expect(screen.queryByTestId("header_aboutDialog")).not.toBeInTheDocument();
    // expect(screen.queryByTestId("header_aboutDialog")).not.toContainElement(null);
    // // expect(setModalOpen).toHaveBeenLastCalledWith(false);
  });
});
