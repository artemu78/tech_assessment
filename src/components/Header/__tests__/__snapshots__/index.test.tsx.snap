// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header render snapshot 1`] = `
<div>
  <header>
    <a
      href="https://war.ukraine.ua/support-ukraine/"
      rel="noreferrer"
      target="_blank"
    >
      <div
        class="praporBlue"
      />
      <div
        class="praporYellow"
      />
    </a>
    <div
      class="titleLine"
    >
      <button
        aria-haspopup="true"
        aria-label="menu"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-18jzufx-MuiButtonBase-root-MuiIconButton-root"
        data-testid="header_menuButton"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <h1>
        Technical Assessment
      </h1>
    </div>
  </header>
</div>
`;

exports[`render Header 1`] = `
<div
  aria-hidden="true"
>
  <header>
    <a
      href="https://war.ukraine.ua/support-ukraine/"
      rel="noreferrer"
      target="_blank"
    >
      <div
        class="praporBlue"
      />
      <div
        class="praporYellow"
      />
    </a>
    <div
      class="titleLine"
    >
      <button
        aria-haspopup="true"
        aria-label="menu"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-18jzufx-MuiButtonBase-root-MuiIconButton-root"
        data-testid="header_menuButton"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <h1>
        Technical Assessment
      </h1>
    </div>
  </header>
</div>
`;
